Rework routing and URL paths

heralden opened the pull request:
This PR is paired with #362. Basically it changes how routes are handled (as well as library used), adding numerous features/fixes along the way to make sure it was a good choice. [=

Here's a copy of the list of goals and related issues:

- [x] Rename reportpage path
- [x] Query string including keyword for *search* path
- [x] Replace *results* path with *list*
- [x] If a list is not in the history, automatically query for it (such as the case when directly opening path `/humanmine/list/PL_Pax6_Targets`) #66
- [x] Make sure report page deep links work from a cold start in a different mine #78
- [x] Deep linking of mines (try to have the mine namespace optional in the URL path, but fallback to specifying `default` if this isn't practical) #352
- [x] Replace secretary+accountant with reitit for reverse-routing (so we can easily rename routes)
- [x] Use anchor elements and href (to allow *open in new window*) #90
- [x] Use HTML5 history API instead of hash history
- [x] Make sure back button works with the new solution

Something I haven't tested is the creation of lists through saving a query, set operations, etc. Although they should *probably* work as `:results/history+` which they rely on still has the same behaviour.

heralden committed:
Replace secretary with reitit using HTML5 history API

Also use reverse routing and href where anchor element is already
available.

heralden committed:
Include name of mine as part of URL pathname

heralden committed:
Merge branch 'dev' into feature/rework-routing

heralden committed:
Improve handling of new HTML5 routing and mine in pathname

heralden committed:
Rework boot-flow and event handlers to support deep links with mines

heralden committed:
Replace uses of on-click with href and add docs to route

heralden committed:
Query string for keyword in search URL path

heralden committed:
Fix routing bug causing lists to not be accessible

heralden committed:
Use /list path and support deep linking of lists

yochannah commented:
@uosl - I'm taking a look at this now, but can I get you to make sure travis passes before we merge?

heralden committed:
Fix linting errors for Travis CI

heralden committed:
Ignore uncaught exception from reitit failing CI

I also removed the other uncaught:exception listener for `assoc`, as it
doesn't seem to apply anymore.

yochannah commented on src/cljs/bluegenes/core.cljs:
```diff
@@ -26,22 +25,24 @@
-          (if (= (first hashless-path) "/")
-            (navigate! hashless-path)
-            (navigate! (str "/" hashless-path))))))
+;; TODO replace or delete this function depending on needs with reitit
```
what are the considerations here with regards to the todo?

heralden replied:
Deep linking seems to be working fine without this, so we can delete this code.

Edit: I can see it's specifically for handling links with the hash router, which we don't use anymore.

yochannah commented on src/cljs/bluegenes/events.cljs:
```diff
@@ -60,24 +60,52 @@
+ :success-init-mine
+ (fn [_ _] {}))
+
+;; This is for setting :current-mine based on the URL path. As opposed to
```
nice comments here!

yochannah commented on src/cljs/bluegenes/route.cljs:
```diff
@@ -0,0 +1,230 @@
+            [reitit.frontend.controllers :as rfc]
+            [reitit.frontend.easy :as rfe]))
+
+;; # Quickstart guide:
```
This is fabulous, thank you!

yochannah commented:
perfect, thanks

On Fri, 19 Jul 2019 at 11:47, Herald <notifications@github.com> wrote:

> *@uosl* commented on this pull request.
> ------------------------------
>
> In src/cljs/bluegenes/core.cljs
> <https://github.com/intermine/bluegenes/pull/372#discussion_r305303862>:
>
> >    (reagent/render [views/main-panel]
>                    (ocall js/document "getElementById" "app")))
>
> -(defn navigate-to-deep-links []
> -  (let [url (oget js/window "location" "hash")
> -        hashless-path (last (clojure.string/split url #"#"))]
> -    (cond (> (count url) 2) ;; if there is more than #/ in the url, navigate there
> -          (if (= (first hashless-path) "/")
> -            (navigate! hashless-path)
> -            (navigate! (str "/" hashless-path))))))
> +;; TODO replace or delete this function depending on needs with reitit
>
> Deep linking seems to be working fine without this, so we can delete this
> code.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/intermine/bluegenes/pull/372?email_source=notifications&email_token=ACGXRDSOHOFFSK7AZQ65YYTQAGLSHA5CNFSM4H4FVPRKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB67H7UI#discussion_r305303862>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACGXRDUILFANZZJATEB3LEDQAGLSHANCNFSM4H4FVPRA>
> .
>

yochannah commented on src/cljs/bluegenes/route.cljs:
```diff
@@ -0,0 +1,230 @@
+                 :params (update params :mine #(or % (:current-mine db)))}}))
+
+(reg-event-db
+ ::navigated
```
could we add a comment about when/why this is used?

heralden replied:
Is it worth it? It's pretty internal to how the router works, so it's not that important to know how it works, and those who do want to know can read it as it's still a pretty short and simple function. [=
I at least can't think of a succinct and useful comment to put there.

heralden replied:
>"you probably won't need to use this as it's an internal helper method" ?

Sure, marking it as internal sounds reasonable.

yochannah commented:
"you probably won't need to use this as it's an internal helper method" ?

On Fri, 19 Jul 2019 at 12:01, Herald <notifications@github.com> wrote:

> *@uosl* commented on this pull request.
> ------------------------------
>
> In src/cljs/bluegenes/route.cljs
> <https://github.com/intermine/bluegenes/pull/372#discussion_r305308441>:
>
> > +;; of all routes, so it gets injected automatically from db if not specified.
> +
> +;; Based on the official reitit frontend-re-frame example: (2019.06.18)
> +;; https://github.com/metosin/reitit/tree/master/examples/frontend-re-frame
> +
> +;;; Events ;;;
> +
> +(reg-event-fx
> + ::navigate
> + (fn [{db :db} [_ route & [params query]]]
> +   {::navigate! {:k route
> +                 :query query
> +                 :params (update params :mine #(or % (:current-mine db)))}}))
> +
> +(reg-event-db
> + ::navigated
>
> Is it worth it? It's pretty internal to how the router works, so it's not
> that important to know how it works, and those who do want to know can read
> it as it's still a pretty short and simple function. [=
> I at least can't think of a succinct and useful comment to put there.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/intermine/bluegenes/pull/372?email_source=notifications&email_token=ACGXRDVQDCNBDRO4GZZD2ZTQAGNHDA5CNFSM4H4FVPRKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB67JLHA#discussion_r305308441>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACGXRDVT46BANOGQXTV47ITQAGNHDANCNFSM4H4FVPRA>
> .
>

yochannah commented on src/cljs/bluegenes/route.cljs:
```diff
@@ -0,0 +1,230 @@
+                 (dispatch [:set-active-panel :reportpage-panel
+                            {:type type :id id :mine mine}
+                            [:load-report mine type id]]))}]}]]])
+;; You can do initialisations by adding a :start function to :controllers.
```
I'm not quite sure what you mean here by teardowns and initialisations here, or why I might want to use them.

heralden replied:
Most of the routes have a `:controllers` vector with a `:start` map. While you can copy-cat the existing `:start` maps when you create new routes, there aren't any using `:stop` (teardowns) currently. I wanted to make sure to write it somewhere so people know they exist as well.

yochannah reviewed (approved):
:medal_sports:
